redis-6.0.9/src/modules/systemtime.c

```diff
@@ -0,0 +1,249 @@
+#include <time.h>
+#include <sys/timerfd.h>
+#include <sys/epoll.h>
+#include <pthread.h>
+#include <unistd.h>
+#include <errno.h>
+#include "../redismodule.h"
+
+static int setup_io_multiplexer();
+static int setup_timer();
+static int start_sys_time_collector_thread();
+
+static int s_epoll_fd = 0;
+static int s_timer_fd = 0;
+static const int NANO_IN_SEC = 1000000000;
+static const int MILLI_IN_SEC = 1000;
+static long s_system_time_usec = 0;
+static double jiffi_in_seconds = 0.0;
+#ifdef __USE_MUTEX
+static pthread_mutex_t lock;
+#endif
+
+/*
+* DRIVENETS.GET.TIME 
+* this is an implementation of an efficient get-system-time mechanism
+* the idea is to use an internal thread that will "collect" the time
+* passing into a static register. once a time sample is required, the
+* command api will sample _the register_ without calling an expensive
+* system call, or probing some hardware clock.
+* 
+* as there is an inherant race on the above register,
+* there are 2 modes with which this can be compiled:
+* 1. -D__USE_MUTEX: uses a pthreads mutex.
+* 2. -D__USE_ATOMICS: uses a gcc builtin atomic operation
+*
+* the user can choose the format of the output:
+* 1. -D__USE_HUMAN_TIME: will format the timestamp to a humanly readable
+*    string. _note_: this option is expensive and "losy" as it looses percision
+*    when it cuts the timestamp down to seconds (the base needed by the
+*    formating functions)
+* 2. none: will return the full timestamp in nanos
+*
+* profiling numbers are given for each build permutation in: profile_results
+*/
+int DrivenetsGetTime_RedisCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+	REDISMODULE_NOT_USED(argv);
+    REDISMODULE_NOT_USED(argc);
+	
+	// dont hold the lock for the ipc
+	long system_time_usec = 0;
+#ifdef __USE_MUTEX
+	pthread_mutex_lock(&lock);
+	system_time_usec = s_system_time_usec;
+	pthread_mutex_unlock(&lock);
+#elif __USE_ATOMICS
+	__atomic_store(&system_time_usec, &s_system_time_usec, __ATOMIC_SEQ_CST);
+#endif
+
+	///// send the result back.
+#ifdef __USE_HUMAN_TIME	
+	time_t time = system_time_usec / NANO_IN_SEC;
+	RedisModule_ReplyWithCString(ctx, asctime(localtime(&time)));
```

Comment:
amazing. you were right about this being an issue.
instead of using the library functions, i used code that "parses" the timestamp just for the fields required.
it improved the performs by a factor of almost 4!!!
cmdstat_drivenets.get.time:calls=11,usec=116,usec_per_call=10.55